extension/background.js

```diff
@@ -1,112 +1,49 @@
+	localStorage.setItem('incHist', JSON.stringify(incHist));
+	localStorage.setItem('incRecent', JSON.stringify(incRecent));
+	});
 
-	var tabInfo = {
-		id: sender.tab.id,
-		url: sender.tab.url,
-		title: sender.tab.title,
-		favIcon: sender.tab.favIconUrl,
-		timestamp: Date()
-	};
+chrome.tabs.onUpdated.addListener(function(tabId, chg, tab){
+	if ( ['chrome://newtab/', 'about:blank'].includes(tab.url) ) return;
 
-	if (localStorage.getItem(windowCount) == null) {
-		count = 1;
-		localStorage.setItem(windowCount, count);
-	}
-
-	if (localStorage.getItem(incogHistory) == null) {
-		var arr = [];
-		localStorage.setItem(incogHistory, JSON.stringify(arr));
-	}
-
-	if (localStorage.getItem(incogRecent) == null) {
-		var arr = [];
-		arr.length = 0;
-		localStorage.setItem(incogRecent, JSON.stringify(arr));
-	}
-
-	tab.storeData(tabInfo);
-
-});
-
-var tab = {
-	storeData: function (tabInfo) {
-		var arr = [];
-		arr = JSON.parse(localStorage.getItem(incogHistory));
-		arr.push(tabInfo);
-
-		localStorage.setItem(incogHistory, JSON.stringify(arr));
-	}
-}
-
-
-function recentRecord(tabId) {
-	var recent = [],
-		history = [],
-		historyLength;
-
-	history = JSON.parse(localStorage.getItem(incogHistory));
-	historyLength = history.length - 1;
-
-	if (tabId != undefined) {
-		var sameTab = [],
-			sameTabLength;
-
-		recent = JSON.parse(localStorage.getItem(incogRecent));
-
-		for (var i = 0; i <= historyLength; i++) {
-			if (tabId == history[i].id)
-				sameTab.push(history[i]);
+	var t, oldU, n;
+	if (t=tabs[tabId]) {
+		oldU=t.url;
+		n=t.n;
 		}
 
-		sameTabLength = sameTab.length;
-		if (sameTabLength > 1)
-			recent.push(sameTab[sameTabLength - 1]);
-		else
-			recent.push(sameTab[0]);
-
-		recent = recent.filter(function (e) { return e != null });
-
-		localStorage.setItem(incogRecent, JSON.stringify(recent));
-	}
-}
+	if ( (chg.status=='loading') || !t) {
+		t=tabs[tabId]={id:tabId, url:tab.url, title:tab.title, timestamp: Date()};
+		if (tab.favIconUrl) t.favIcon=tab.favIconUrl;
+		if (oldU != t.url) {
+			t.n=incHist.length;
+			incHist.push(t);
+			}
+		else if (n !== undefined) {
+			// re-linking
+			t.n=n;
+			incHist[n]=t;
+			}
+		}
+	else if ( (chg.status=='complete') || chg.title || chg.favIconUrl) {
+		t.title=tab.title;
+		if (tab.favIconUrl) t.favIcon=tab.favIconUrl;
+		}
+	});
 
+chrome.tabs.onReplaced.addListener(function(newId, oldId){
+	console.info('replace tab', newId, oldId);
+	if (tabs[oldId]) {
+		tabs[newId]=tabs[oldId];
+		tabs[newId].id=newId;
+		}
+	});
 
 chrome.tabs.onRemoved.addListener(function (tab) {
-	recentRecord(tab);
-});
-
-chrome.windows.onCreated.addListener(function (window) {
-	if (localStorage.getItem(windowCount) == null) {
-		count = 1;
-		localStorage.setItem(windowCount, count);
-	}
-	else {
-		var count = parseInt(localStorage.getItem(windowCount));
-		count++;
-		localStorage.setItem(windowCount, count);
-	}
-});
-
-chrome.windows.onRemoved.addListener(function () {

```

Comment:
We need this functionality to maintain our feature where the history gets deleted automatically when all incognito windows are closed.

Reply:
No, because when the last incognito window is closed, the background page/script is destroyed by chrome (remember; we are in split mode here). There is nothing left to update.
You might think otherwise because when you're debugging, and have the inco-background devtools window opened, there is still an incognito window opened: the devtools window.

Try this:
- open an incognito window
- open `chrome://extensions/?id=WithTheIdOfTheExtension` and keep it opened.
you'll see:
```
Inspect views
background page (Inactive)
background page (Incognito)
```
- open the incognito background page view
- close the incognito window. Keep the devtools open.
- refresh the `chrome://extensions/?id=` , you'll still see the 2 views
- close the devtools window, refresh `chrome://extensions/?id=`  ,  `background page (Incognito)` has disappeared

Reply:
Right, it is working correctly as you explained.

I thought Chrome would destroy background page/script after a certain period of time, but it destroys page/script immediately.